File: 76.minimum-window-substring.py
#
# @lc app=leetcode id=76 lang=python3
#
# [76] Minimum Window Substring
#

# @lc code=start


class Solution:
    def minWindow(self, s: str, t: str) -> str:
        sn, tn = len(s), len(t)
        if tn > sn:
            return ""
        idx = lambda x: ord(x) - ord("A")
        res = (0, float("inf"))
        need = [0] * 58
        
        for c in map(idx, t):
            need[c] += 1
        l, needcnt = 0, len(t)

        for r, c in ((r, idx(c)) for r, c in enumerate(s)):
            if need[c] > 0: # c is in t, 1 less char to match
                needcnt -= 1
            need[c] -= 1

            if needcnt > 0: 
                continue
            while need[i := idx(s[l])] < 0: # remove preceding chars not needed
                need[i] += 1
                l += 1
            if r - l < res[1] - res[0]: 
                res = (l, r)

            need[idx(s[l])] += 1 # remove the first char in window
            needcnt += 1
            l += 1
        return s[res[0] : res[1] + 1] if res[1] != float("inf") else ""




# @lc code=end

File: 76.minimum-window-substring.rs
/*
 * @lc app=leetcode id=76 lang=rust
 *
 * [76] Minimum Window Substring
 */

// @lc code=start
impl Solution {
    pub fn min_window(s: String, t: String) -> String {
        let (sn, tn) = (s.len(), t.len());
        let (sbytes, tbytes) = (s.as_bytes(), t.as_bytes());
        if sn < tn {
            return "".to_owned();
        }
        let idx = |c: &u8| -> usize { (c - b'A') as usize };
        let mut need = [0; 58];
        let mut res = (0, usize::max_value());

        for c in tbytes.iter().map(|c| idx(c)) {
            need[c] += 1;
        }
        let (mut l, mut needcnt) = (0, tn);

        for (r, c) in sbytes.iter().map(|c| idx(c)).enumerate() {
            if need[c] > 0 {
                needcnt -= 1;
            }
            need[c] -= 1;
            if needcnt > 0 {
                continue;
            }
            while need[idx(&sbytes[l])] < 0 {
                need[idx(&sbytes[l])] += 1;
                l += 1;
            }
            if r - l < res.1 - res.0 {
                res = (l, r);
            }
            need[idx(&sbytes[l])] += 1;
            needcnt += 1;
            l += 1;
        }

        if res.1 == usize::max_value() {
            return "".to_owned();
        }
        s[res.0..=res.1].to_owned()
    }
}
// @lc code=end
